### ISSUE TITLE
Lua web sockets library

### ISSUE THREAD

**Issue by VoidAeternum** [Facepunch/garrysmod-requests#1063]
Hello! I worked with lua, and got a trouble sending messages from Node.JS to GMod and backwards. Web sockets is a great utility for modders. That would add an opportunity for integrating gmod with such things as server web chat, server-to-server communication and etc. Example:

**Example:**

```Lua
local ws = socket.Open(port)

socket.Receive(ws, "connection", function(...)
    ...
end)

socket.Receive(ws, "close", function(...)
    ...
end)

```

```Lua
local ws = socket.Connect("localhost:3000")

function ws:OnOpen()
    MsgN("Connection opened!")
end

function ws:OnClose() ... end

function ws:OnMessage(data?) ... end
```

```Lua
ws:Send("{hello = 'world'}")
```

**Comment by Kefta** [comment]
A default socket library would be nice, but you can already use a socket module like gmsocket or bromsock

**Comment by VoidAeternum** [comment]
Using dll's is very inconvinient, and native library would be a great tool. With native way of that you can be sure that every server can support it. For example, server-to-server wiremod antenna with a node server in the center of everything.

**Comment by sannysc** [comment]
You could say the same thing about mysql dlls. Tons of addons already use them, and like @Kefta said, it'd be nice to have built in features like these, but we also have to take into consideration the fact that these things take time and effort to make (and maintain), and we're already getting along just fine using dlls.

**Comment by edgarasf123** [comment]
Default socket library might cause security issues, especially on shared servers.

**Comment by neico** [comment]
^

Which is why I'd opt for a websocket library instead, you can't fuck things up as much but still can do what most others do already with sockets (IRC Relay to name one)

**Comment by VaasKahnGrim** [comment]
This would also help alot if it were usable both clientside and serverside.

**Comment by VaasKahnGrim** [comment]
another 2 years and this shit still hasn't been looked back at. What a shame

**Comment by GitSparTV** [comment]
Awesomium and Chromium had websockets as well but this led to discord protocol abusing. Having separate api for websockets seen unreal for now

**Comment by GitSparTV** [comment]
Unless Gmod rethinking features removing and look forward server bans

**Comment by VaasKahnGrim** [comment]
> Awesomium and Chromium had websockets as well but this led to discord protocol abusing. Having separate api for websockets seen unreal for now

yea but that was discord's fault for not taking security seriously. Also I'm pretty sure that the same thing that currently stops http.Post/Fetch from connecting to local addresses could also be used for preventing sockets from fucking with discord clients aswell. Like I don't see why its too much of an issue with that.

**Comment by VaasKahnGrim** [comment]
Like it would be nice to not have to funnel traffic through SRCDS to my own software and instead go directly between the client and my own stuff

**Comment by GitSparTV** [comment]
Discord doesn't have security issues, it's API for games and interactions, the same thing discord uses for resolving invites from browser in app

**Comment by VaasKahnGrim** [comment]
So its a feature not a bug, but either way the stuff for http.Post/Fetch preventing local ips would have solved the socket issue

**Comment by GitSparTV** [comment]
local ips are blocked for the same reason

**Comment by VaasKahnGrim** [comment]
yeah exactly, the same thing could just make sockets a safe and very useful

**Comment by robotboy655** [comment]
I couldn't find a way to make that happen for websockets for either Awesmoim or CEF.

**Comment by VaasKahnGrim** [comment]
> I couldn't find a way to make that happen for websockets for either Awesmoim or CEF.

I don't see how it would be that hard to do it. a connect() function in C++ could just check what string is being given to it from lua and see if it matches a string in a blacklist. and siad blacklist is just a few strings like localhost, 192.168.X.X, 127.0.0.0

For the blacklist with IPs you could just focus on the first 2 numbers in the ip using something like
```c++
std::string ip_start = ip.substr(0, s.find("."))

if (ip_start == "127" || ip_start == "192" || ip == "localhost") {
	//deny socket connection before its even started
}
```

**Comment by GitSparTV** [comment]
Local ip ranges are wider, but you are right the ip filter should be implemented properly

**Comment by VaasKahnGrim** [comment]
yea but I didn't feel like going with all the reserved IPs. But substr() would make it alot easier than having to check the entire IP

https://en.wikipedia.org/wiki/Reserved_IP_addresses

**Comment by VaasKahnGrim** [comment]
> I couldn't find a way to make that happen for websockets for either Awesmoim or CEF.

Since you mentioned it earlier I felt I'd spend a few mins typing this thing out. Pretty much covers every reserved IP and Private Network IP listed(including localhost aswell as any IP on the same network) So pretty much any ip fed into this that Lua should never try to contact with sockets will be blocked but everything else will pass right on through.

It should be all set for you to use, just modify where CEF and Awesomium make socket connections from and have it pass IPs through this function first. Alternatively a native means of using sockets that doesn't require CEF or Awesomium would be better. Infact I'm probably going to also use this function myself for other things. But for now its 1am as of posting so I'm going to sleep.

ConnectionIsBlacklisted(ip) - returns true if blocked, false if allowed
https://pastebin.com/70Fpixji

**Comment by robotboy655** [comment]
> just modify where CEF and Awesomium make socket connections

This is what I meant when I said I couldn't find a way. There are no APIs for this in either CEF or Awesomium as far as I can tell.

**Comment by VaasKahnGrim** [comment]
In that case I guess a dedicated means of handling socket connections that doesn't involve CEF or Awesomium is the way to go. Which would actually be alot better as we wouldn't have to deal with RunJavascript anymore for it. It would probably also be easier to maintain aswell as get working with that filter I sent(or what ever filter you decide might be better I guess)

**Comment by VaasKahnGrim** [comment]
this probably isn't the best way but I just did a quick addition to the function I sent. This being to handle IPs that are converted to HEX, Octal or Decimal. Also should handle attempts to bypass the filter with a DNS entry. People won't be able to connect to URLs and will have to go specifically to an IP but I feel thats a small price to pay.

```c++
//Replace if statement at top of ConnectionIsBlacklisted()
if (hostname.length <= 7 || !hostname.find('.') || !hostname.find('/')) {
	//Prevent attempt to connect to an ip thats too short or invalid
	//Also blocks attempted bypasses using IPs as Hex, Octal or decimal
	//Also blocks attempts to connect to an ip/hostname using / in it(typically you would only connect to the hostname:port you wouldn't connect using a file path like that until AFTER being connected) This also would block attempts to connect to urls that try and bypass the filter using DNS
	return true;
}
```

**Comment by thegrb93** [comment]
Web sockets are not regular sockets. Use a binary module if you want regular sockets.

**Comment by VaasKahnGrim** [comment]
sockets are sockets as far as I care. either websockets or sockets being available for use would work fine

**Comment by Kefta** [comment]
They're vastly different.

**Comment by VaasKahnGrim** [comment]
> They're vastly different.

They both still accomplish the same thing tho.
